Contributing.md

```diff
@@ -1,30 +1,471 @@
+managed by the CI bots. E.g.
+
+* `P-waiting_on_author` - The approval process is paused pending action from the PR author.
+* `P-waiting_on_reviewer` - The approval process is paused pending action from a reviewer.
+* `P-approved` - The PR has been approved and will be merged soon. 
+* `P-queued_for_merge` - The PR is being merged by the CI process.
+
+### Experience level (E)
+
+These labels indicate the level or type of experience needed to address. Examples:
+
+* `E-good_first_issue`
+* `E-cryptographer`
+* `E-medium`
+* `E-expert`
+
+
+[git commit farming]: https://www.tari.com/git-farming/ "The Tari Blog: Git farming or, How to get your PRs merged into Tari"
+
+# Automated CI checks
+
+The Tari CI process carries out several checks that must pass before a PR will be merged, or even reviewed:
+
+* PR title
+* Licence information
+* Formatting
+* Coding style
+* Unit tests
+
+## PR title
+
+Commit messages, or more specifically, PR titles, need to conform to the 
+[Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) conventions.
+
+If you fall afoul of this, simply edit your PR title, and the test will automatically run again.
+
+
+## License information
+
+All new source files MUST have relevant license information at the top of the file. Use the BSD-3 license for all 
+Tari code.
+
+Scripts, HTML, CSS, binary files, configuration files, and other non-source types are not subject to this
+requirement. See the [license check script] for the complete list of extensions that are ignored for this check.
+
+The following text is recommended:
+
+```text
+// Copyright 2022 The Tari Project
+// SPDX-License-Identifier: BSD-3-Clause
+```
+
+You can run the [license check script] to help identify any files that are missing 
+license info.
+
+[license check script]: ./scripts/file_license_check.sh
+
+## Formatting
 
 PRs are checked for formatting using the command:
+
 ```
-TOOLCHAIN_VERSION=nightly-2019-07-15
-rustup component add --toolchain $TOOLCHAIN_VERSION rustfmt
 cargo fmt --all -- --check
 ```
 
-You can automatically format the code using the command:
+If your tests fail here, simply run
+
 ```
 cargo fmt --all
 ```
 
-The toolchain version may change from time to time, so check in `.circlecit/config.yml` for the latest toolchain version
+If this does not work:
+
+* Is rust fmt installed? `rustup component add rustfmt` 
+* It is installed, but the check still fails? Check that your rust version matches the one in our CI toolchain. You 
+  can inspect this in the `env.toolchain` field in the [Github Action CI workflow](.github/workflows/ci.yml)
+
+# Code style
+
+Tari uses Clippy to encourage consistent and idiomatic Rust code.
+
+Clippy does not support project-wide Clippy configuration files at the moment. Tari uses `cargo-lints` instead and 
+defines global linting rules in `lints.toml`.
+
+The CI enforces these lints, so check your code with:
+
+```
+cargo lint clippy --all-targets
```

Comment:
```suggestion
cargo lint clippy --all-targets --all-features
```

Reply:
Correct!

Reply:
That's not what currently runs in CI. I'm happy to add the `--all-features` flag, but we must update the CI to reflect this.

Reply:
Added in #4165